docs/architecture/adr-017-single-binary-upgrade.md

```diff
@@ -0,0 +1,141 @@
+
+- Will celestia-node need to pass a block height to celestia-app? Header contains version.app (state machine version) and version.block (consensus version)
+
+- Is it important to support rolling back versions? Involves upgrading (again) through the faulty version
+
+- Since if conditional logic uses one Go version, all code will be compiled with this one Go version. If behavior changes across Go versions, we have to debug why.
+
+- Option: depend on multiple major versions of a dependency. What is the relationship between if/else conditional statements and Cosmos SDK upstream changes?
+
+- How to do this in CometBFT?
+  - Always use latest p2p version
+  - Version.block in header should be used to version changes to Tendermint / CometBFT
+
+- Block sync needs to be aware of changes across block heights
+
+### Implementation proposal
+
+There are multiple ways of doing it. The app version can be found in the SDK context.
+So, in the `PrepareProposal` method, we can get the app version [from the context](https://github.com/celestiaorg/celestia-app/blob/main/app/prepare_proposal.go#L28) and use it as the version indicator.
+
+#### Option 1: Passing the app version throughout all the required functions
+
+In this case we need to pass either sdk context or the app version itself to all the functions and objects that we think they might be affected by a version change in future.
+So for those functions that do not have the ctx already, we need to add an extra parameter to pass it.
+
+#### Option 2: Define a version package and call a global function to get the version
+
+We can define a package e.g. `version` that has a setter and getter. And the getter can be called by anyone throughout the application.
+
+```go
+var appVersion string
+
+func SetVersion( newVer string) error {
+  // Check if the caller is eligible to call this.
+  appVersion = newVer
+}
+
+func GetVersion() string {
+  return appVersion
+}
+
+```
+
+Then we call the setter function in `ProcessProposal` where it reads it from the block header.
+The getter function will be called everywhere e.g. [mergeMaps](https://github.com/celestiaorg/celestia-app/blob/f1dec1014a7159c0f0b213182aff4793163e9732/pkg/shares/share_splitting.go#L159)
+
+```go
+func mergeMaps(mapOne, mapTwo map[coretypes.TxKey]ShareRange) map[coretypes.TxKey]ShareRange {
+	merged := make(map[coretypes.TxKey]ShareRange, len(mapOne)+len(mapTwo))
+
+  switch version.GetVersion() {
+
+    case version.V0:
+      maps.Copy(merged, mapOne)
+      maps.Copy(merged, mapTwo)
+
+    case version.V1:
+      maps.Copy(merged, mapTwo)
+      maps.Copy(merged, mapOne)
+
+    default:
+        panic(version.ErrVersionNotFound)
+  }
+  return merged
+}
+```
+
+This option has an advantage over the option #1 as we do not need to pass an extra parameter everywhere in the code and it has less maintenance burden to modify the version logic.
```

Comment:
A counterpoint is that option 2 makes it more difficult to unit test because pure functions are easier to test than impure functions.

Implementations in option 2 rely on a global state. It is likely still possible to unit test by mocking the version module with a test version module but I think it makes sense to call this out as a downside.

Related question: how do we ensure all necessary packages that want to read the global version will invoke it _after_ it has been set?